pkg/engine/events.go

```diff
@@ -188,13 +188,62 @@ func makeEventEmitter(events chan<- Event, update UpdateInfo) (eventEmitter, err
 
 	logging.AddGlobalFilter(logging.CreateFilter(secrets, "[secret]"))
 
+	// Instead of sending to the source channel directly, buffer events to account for slow receivers.
+	//
+	// Buffering is done by a goroutine that concurrently receives from the senders and attempts to send events to the
+	// receiver. Events that are received while waiting for the receiver to catch up are buffered in a slice.
+	//
+	// We do not use a buffered channel because it is empirically less likely that the goroutine reading from a
+	// buffered channel will be scheduled when new data is placed in the channel.
+	buffer, done := make(chan Event), make(chan bool)
+	go func() {
+		defer close(done)
+
+		var queue []Event
+		for {
+			contract.Assert(buffer != nil)
+
+			e, ok := <-buffer
+			if !ok {
+				return
+			}
+			queue = append(queue, e)
+
+			// While there are events in the queue, attempt to send them to the waiting receiver. If the receiver is
+			// blocked and an event is received from the event senders, stick that event in the queue.
+			for len(queue) > 0 {
+				select {
+				case e, ok := <-buffer:
+					if !ok {
+						// If the event source has been closed, flush the queue.
+						for _, e := range queue {
+							events <- e
+							if e.Type == CancelEvent {
+								return
```

Comment:
I'm not sure the exact contract around `e.Type` - but what's the benefit to returning here instead of allowing the queue to flush?  

More generally - other than this `CancelEvent` code, this looks like a pure channel abstraction that doesn't interfere with the semantics of the participants on either side.  Is there a reason this knowledge of the contract of `CancelEvent` needs to be handled in this layer?

Reply:
That's a fair point. I'll try changing this.